### PR TITLE
Fix intrinsics being considered as overload candidates in namespace-qualified calls

### LINKED ISSUE #1884: namespaces and normalize function errors
Hello we have discovered issues with namespaces:

```
Texture2D<float4> texture;
RWTexture2D<float3> textureOut : register(u0);

namespace safe1
{
  static const float SAFE_EPSILON = 2.4414e-4;
  float3 normalize(float3 _x) { return _x * rsqrt(max(SAFE_EPSILON, dot(_x, _x))); }
  float2 normalize(float2 _x) { return _x * rsqrt(max(SAFE_EPSILON, dot(_x, _x))); }
  float3 Normalize(float3 _x) { return _x * rsqrt(max(SAFE_EPSILON, dot(_x, _x))); }
}

[numthreads(16,16,1)]
void CS_Main(uint3 groupId : SV_GroupID,
    uint3 dispatchThreadId : SV_DispatchThreadID,
    uint3 groupThreadId : SV_GroupThreadID
)
{
	float4 l = texture.Load(0); 
	float3 n = safe1::normalize(float3(l.x,l.y,l.z));
	textureOut[uint2(0,0)] =float3(n);
}

```
This code outputs equivalent to a call to default hlsl normalize when you have two normalize functions in the namespace as pointed out.

If you have only 1 it calls it correctly.

Here is the command line

`dxc.exe shader_preprocessed.hlsl /O3 /E CS_Main /T cs_6_0 -Fc shader_preprocessed.asm`

To recap.

safe1::normalize calls is equivalent to ::normalize
safe1::Normalize is correct.

### PR DESCRIPTION
**Pull request by MrTrillian**
Intrinsics should never be candidates for qualified calls, except global qualifier.

Fixes #1884

### CODE REVIEW

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 1]
```diff
@@ -4069,6 +4069,12 @@ class HLSLExternalSource : public ExternalSemaSource {
   {
     DXASSERT_NOMSG(ULE != nullptr);
 
+    // Intrinsics live in the global namespace, so references to their names
+    // should be either unqualified or '::'-prefixed.
+    if (ULE->getQualifier() && ULE->getQualifier()->getKind() != NestedNameSpecifier::Global) {
```
Comment (vcsharma) [inline tools/clang/lib/Sema/SemaHLSL.cpp]: Does fxc allow overloading intrinsic from non-global namespace? Also, what happens if we try to overload an intrinsic from the global namespace? I think fxc allows that? If yes, should we have an xfailing test to cover that scenario and address it later?

File: tools/clang/test/CodeGenHLSL/declarations/functions/overloading/namespace_qualified_no_intrinsic_candidate.hlsl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+// Regression test for GitHub #1884, where intrinsics were considered
+// valid overload candidates for overloaded functions of the same name in a namespace
+
+// CHECK: call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 42)
+
+namespace foo
+{
+    int abs(int2 x) { return 42; }
+    int abs() { return 0; }
```
Comment (vcsharma) [inline tools/clang/test/CodeGenHLSL/declarations/functions/overloading/namespace_qualified_no_intrinsic_candidate.hlsl]: if you defined this function, then better to call this one as well in main and add a CHECK for it.

Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/declarations/functions/overloading/namespace_qualified_no_intrinsic_candidate.hlsl]: Hmm, what more would that be testing?

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 1 above)
Comment (MrTrillian) [reply tools/clang/lib/Sema/SemaHLSL.cpp]: Added more tests to cover those cases, and filed #1887

File: tools/clang/test/CodeGenHLSL/declarations/functions/overloading/namespace_qualified_no_intrinsic_candidate.hlsl (hunk 2 above)
Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/declarations/functions/overloading/namespace_qualified_no_intrinsic_candidate.hlsl]: I'm confused, main is what emits storeOutput, not the overloads...? There is certainly a need for overload tests, but that's a task of its own.

Comment (vcsharma) [reply tools/clang/test/CodeGenHLSL/declarations/functions/overloading/namespace_qualified_no_intrinsic_candidate.hlsl]: Ok. Btw, this is what I meant:

```
// RUN: %dxc -E main -T vs_6_0 %s | FileCheck %s

// Regression test for GitHub #1884, where intrinsics were considered
// valid overload candidates for overloaded functions of the same name in a namespace

// CHECK: call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 42)
// CHECK: call void @dx.op.storeOutput.i32(i32 5, i32 0, i32 0, i8 0, i32 0)

namespace foo
{
    int abs(int2 x) { return 42; }
    int abs() { return 0; }
}

int2 main() : OUT
{
    // This should not consider the abs(int) intrinsic from the global namespace,
    // regardless of the fact that it is a better match for the arguments.
    return int2(foo::abs(0), foo::abs());
}
```

Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/declarations/functions/overloading/namespace_qualified_no_intrinsic_candidate.hlsl]: My last commit includes this.

**Pull request merged**